keras_nlp/src/layers/modeling/reversible_embedding.py

```diff
@@ -165,3 +180,105 @@ def compute_output_spec(self, inputs, reverse=False):
         else:
             output_shape += [self.output_dim]
         return keras.KerasTensor(output_shape, dtype=self.dtype)
+
+    # Quantization-related (int8) methods
+
+    def quantized_call(self, inputs, reverse=False):
+        # TODO (hongyu): This function could be removed once we add `*args` and
+        # `**kwargs` for `Embedding.quantized_call`
+        if self.quantization_mode == "int8":
+            return self._int8_call(inputs, reverse=reverse)
+        else:
+            self._quantization_mode_error(self.quantization_mode)
+
+    def _int8_build(
+        self,
+        embeddings_initializer="zeros",
+        embeddings_scale_initializer="ones",
+        reverse_embeddings_initializer="zeros",
+        reverse_embeddings_scale_initializer="ones",
+    ):
+        super()._int8_build(
+            embeddings_initializer, embeddings_scale_initializer
+        )
+        self.inputs_quantizer = keras.quantizers.AbsMaxQuantizer(axis=-1)
+        if not self.tie_weights:
+            self.reverse_embeddings = self.add_weight(
+                name="reverse_embeddings",
+                shape=(self.output_dim, self.input_dim),
+                initializer=reverse_embeddings_initializer,
+                dtype="int8",
+                trainable=False,
+            )
+            self.reverse_embeddings_scale = self.add_weight(
+                name="reverse_embeddings_scale",
+                shape=(self.input_dim,),
+                initializer=reverse_embeddings_scale_initializer,
+                trainable=False,
+            )
+
+    def _int8_call(self, inputs, reverse=False):
+        if reverse:
+            if self.tie_weights:
+                kernel = ops.transpose(self._embeddings)
+                scale = ops.transpose(self.embeddings_scale)
+            else:
+                kernel = self.reverse_embeddings
+                scale = self.reverse_embeddings_scale
+            inputs, inputs_scale = self.inputs_quantizer(inputs)
+            outputs = ops.matmul(inputs, kernel)
+            # De-scale outputs
+            outputs = ops.cast(outputs, self.compute_dtype)
+            outputs = ops.divide(outputs, ops.multiply(inputs_scale, scale))
+            return outputs
+
+        return super()._int8_call(inputs)
+
+    def quantize(self, mode):
```

Comment:
Could we chain to super here to keep most of the logic? and just handle the `if mode == "int8" and not self.tie_weights` case below? Would be great to keep as much logic on the super class as we can.

Reply:
I'm afraid not.
The raising of `NotImplementedError` in `keras.layers.Embedding` is intentional and inevitable without modifying the Keras code.
The idea is to prevent undefined behavior when users call `Model.quantize`.

I can introduce an argument like `type_check=True` in `keras.layers.Embedding` to support `super` in the future.
However, for now, we can only implement quantize from scratch.

Reply:
I see, thanks for the explainer.

Not to solve on this PR, but I wonder if we can make the contract between Keras and downstream here more public and minimal. I see `_int_8_call()`, `_int_8_build()`, `_quantization_mode_error()`, `_tracker`, and `_untrack_variable()` all used here. That's a pretty significant level of private usage, which could easily break.

Separate question, will this work with older version of Keras 3? Or are there small changes we could make so we don't break older versions?

Reply:
I agree that these methods are too verbose for downstream project. I will try to simplify the contract in the future, but currently, I don't have a good idea for it.

> will this work with older version of Keras 3?
I haven't check the compatibility. My rough guess is that users will need `keras>=3.4.0` due to the introduction of `DTypePolicyMap`